src/core/algorithms/od/fastod/data_frame.cpp

```diff
@@ -0,0 +1,196 @@
+                   ExtractRangesFromColumn);
+
+    for (size_t column = 0; column < cols_num; ++column) {
+        size_t tuple_count = data_.at(column).size();
+
+        std::vector<size_t> curr_column_item_placement;
+        curr_column_item_placement.reserve(tuple_count);
+
+        for (size_t i = 0; i < tuple_count; ++i) {
+            std::optional<size_t> range_index = FindRangeIndexByItem(i, data_ranges_[column]);
+            assert(range_index.has_value());
+
+            curr_column_item_placement.push_back(range_index.value());
+        }
+
+        range_item_placement_.push_back(std::move(curr_column_item_placement));
+    }
+
+    RecognizeAttributesWithRanges();
+}
+
+int DataFrame::GetValue(int tuple_index, AttributeSet::size_type attribute_index) const {
+    return data_[attribute_index][tuple_index];
+}
+
+std::vector<std::vector<DataFrame::value_indexes_t>> const& DataFrame::GetDataRanges() const {
+    return data_ranges_;
+}
+
+size_t DataFrame::GetRangeIndexByItem(size_t item, AttributeSet::size_type attribute) const {
+    return range_item_placement_.at(attribute).at(item);
+}
+
+AttributeSet const& DataFrame::GetAttributesWithRanges() const {
+    return attrs_with_ranges_;
+}
+
+AttributeSet::size_type DataFrame::GetColumnCount() const {
+    return data_.size();
+}
+
+std::size_t DataFrame::GetTupleCount() const {
+    return data_.size() > 0 ? data_.at(0).size() : 0;
+}
+
+bool DataFrame::IsAttributesMostlyRangeBased(AttributeSet attributes) const {
+    if (!attributes.any()) {
+        return false;
+    }
+
+    AttributeSet remaining_attrs = intersect(attrs_with_ranges_, attributes);
+
+    AttributeSet::size_type attrs_count = attributes.count();
+    AttributeSet::size_type remaining_attrs_count = remaining_attrs.count();
+
+    double const accept_range_based_partition_factor = 0.5;
+
+    return (double)remaining_attrs_count / attrs_count >= accept_range_based_partition_factor;
+}
+
+DataFrame DataFrame::FromCsv(std::filesystem::path const& path, char separator, bool has_header,
+                             config::EqNullsType is_null_equal_null) {
+    std::shared_ptr<CSVParser> parser = std::make_shared<CSVParser>(path, separator, has_header);
+    return FromInputTable(parser, is_null_equal_null);
+}
+
+DataFrame DataFrame::FromInputTable(config::InputTable input_table,
+                                    config::EqNullsType is_null_equal_null) {
+    std::vector<model::TypedColumnData> columns_data =
+            model::CreateTypedColumnData(*input_table, is_null_equal_null);
+
+    return DataFrame(std::move(columns_data));
+}
+
+void DataFrame::RecognizeAttributesWithRanges() {
+    double const accept_factor = 0.001;
+
+    for (size_t i = 0; i < data_ranges_.size(); ++i) {
+        size_t items_count = data_[i].size();
+        size_t ranges_count = data_ranges_[i].size();
+
+        if ((double)ranges_count / items_count >= accept_factor) {
```

Comment:
Could you explain the purpose of `accept_factor`, please?

Reply:
We must determine which attributes are range-based (that is, they contain a large number of values that are combined into ranges). While the algorithm is running, this information is used to select the appropriate partition type. This variable serves as the point at which we consider the attribute to be range-based.